toolchain/base/value_store_chunk.h

```diff
@@ -0,0 +1,177 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_BASE_VALUE_STORE_CHUNK_H_
+#define CARBON_TOOLCHAIN_BASE_VALUE_STORE_CHUNK_H_
+
+#include <bit>
+#include <cstddef>
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+#include "common/check.h"
+#include "llvm/ADT/StringRef.h"
+#include "toolchain/base/mem_usage.h"
+
+namespace Carbon::Internal {
+
+// The max size of each chunk allocation for ValueStore. This is based on TLB
+// page sizes for the target platform.
+//
+// See https://docs.kernel.org/admin-guide/mm/hugetlbpage.html
+template <class IdT>
+  requires requires { typename IdT::ValueType; }
+static constexpr auto PlatformChunkMaxAllocationBytes() -> int {
+#if !defined(NDEBUG) || defined(FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION)
+  // Use a small size in debug/fuzzer builds to ensure multiple chunks get used.
+  return sizeof(typename IdT::ValueType) * 5;
+#else
+  // TODO: Should ia64 use 1M or 4M? Should Windows and Mac use different sizes?
+
+  // Linux x64 uses 2M, as x64 CPUs support 4K and 2M page sizes.
+  return 2 * 1024 * 1024;
+#endif
+}
+
+// The number of elements stored in each chunk allocation.
+//
+// The number must be a power of two so that that there are no unused values in
+// bits indexing into the allocation.
+template <class IdT>
+  requires requires { typename IdT::ValueType; }
+static constexpr auto PlatformChunkCapacitySize() -> int {
+  constexpr auto MaxElements =
+      PlatformChunkMaxAllocationBytes<IdT>() / sizeof(typename IdT::ValueType);
+  constexpr auto Pow2MaxElements = std::bit_ceil(MaxElements);
+  if (Pow2MaxElements > MaxElements) {
+    return Pow2MaxElements / 2;
+  }
+  return Pow2MaxElements;
+}
+
+// The number of bits needed to index each element in a chunk allocation.
+template <class IdT>
+  requires requires { typename IdT::ValueType; }
+static constexpr auto PlatformChunkCapacityBits() -> int {
```

Comment:
What do you think of `Index` here, since it's the bits needed for storing an index value? i.e.:

```suggestion
static constexpr auto PlatformChunkIndexBits() -> int {
```

Reply:
Sure, no strong feelings either way, they both work for me. Done.